src/training/graph_group_async.cpp

```diff
@@ -207,30 +208,48 @@ void AsyncGraphGroup::execute(Ptr<data::Batch> batch) {
       gradients = accGradients;
 
       // Keep track of how many words we've calculated the error from
-      num_seen_words += batch_words;
+      num_seen_words += batch->words();
+      num_seen_trg += batch->wordsTrg();
+      num_seen_sentences += batch->size();
     } else {
       gradients = graph->params()->grads();
-      num_seen_words = batch_words;
+      num_seen_trg = batch->wordsTrg();
     }
 
     t++;
 
     if(t % tau_ == 0) {
-      pushGradients(gradients, num_seen_words, t_id);
-      // Reset the counter of seen words after gradient update
-      num_seen_words = 0;
-
+      pushGradients(gradients, num_seen_trg, t_id);
+      // Reset the counter of seen target words after gradient update
+      num_seen_trg = 0;
       if(tau_ > 1)
         gradients->set(0);
     }
 
-    if(scheduler_) {
+    if(t % tau_ == 0 && scheduler_) {
       std::unique_lock<std::mutex> lock(schedulerMutex_);
 
       // Wait until the thread that wants to do validation is finished.
       pool_->wait_for_one(lock);
 
-      scheduler_->update(cost, batch);
+      if (options_->get<std::string>("cost-type") != "ce-sum")
+        cost /= tau_;
+
+      if (tau_ > 1) {
+        std::vector<size_t> fakeLength = {1, 1}; 
+        auto fb = data::CorpusBatch::fakeBatch(fakeLength,
```

Comment:
This seems hacky?

Reply:
@kpu it is. Because we need to inform the scheduler the accumulated batch for statistic purpose (wps, when to start a new epochs, etc2). This is the "best" way I can find to create a fake batch for that purpose.

Reply:
Urgh. Hate this. Why update the scheduler at all?